src/Cake.Common/Build/TeamCity/TeamCityProvider.cs

```diff
@@ -0,0 +1,205 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Cake.Core;
+using Cake.Core.IO;
+
+namespace Cake.Common.Build.TeamCity
+{
+    /// <summary>
+    /// Responsible for communicating with TeamCity.
+    /// </summary>
+    public sealed class TeamCityProvider
+    {
+        private readonly ICakeEnvironment _environment;
+
+        internal TeamCityProvider(ICakeEnvironment environment)
+        {
+            _environment = environment;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the current build is running on TeamCity.
+        /// </summary>
+        /// <value>
+        /// <c>true</c> if the current build is running on TeamCity; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsRunningOnTeamCity
+        {
+            get { return !string.IsNullOrWhiteSpace(_environment.GetEnvironmentVariable("TEAMCITY_VERSION")); }
+        }
+
+        /// <summary>
+        /// Write a progress message to the TeamCity build log.
+        /// </summary>
+        /// <param name="message">Build log message.</param>
+        public static void ProgressMessage(string message)
+        {
+            WriteServiceMessage("progressMessage", message);
+        }
+
+        /// <summary>
+        /// Write a progressStart message to the TeamCity build log.
+        /// </summary>
+        /// <param name="message">Build log message.</param>
+        public static void StartProgress(string message)
```

Comment:
Would be nice to return an IDisposable for the different Start blocks that would call the corresponding End method on Dispose. This would allow for code like this:

using(var scope = TeamCity.WriteProgress("Hello World"))
{
    // Do stuff.
}